File: register-elevenlabs-tools.mjs
#!/usr/bin/env node
// register-elevenlabs-tools.mjs
//
// One-shot script to bulk-register every client tool from
// elevenlabs-client-tools.json with the ElevenLabs Convai API and attach the
// resulting tool IDs to the AI Mix Masters agent's prompt.tool_ids list.
//
// Why: pasting 25 tools one by one in the dashboard form is brutal.
//
// What it does:
//   1. Reads elevenlabs-client-tools.json (sibling file in this folder)
//   2. POSTs each entry to https://api.elevenlabs.io/v1/convai/tools as a
//      `client` tool with `expects_response: true` and a 20s timeout
//   3. Collects the returned tool IDs
//   4. PATCHes the agent (via /v1/convai/agents/{agentId}) so its
//      conversation_config.agent.prompt.tool_ids contains the full list
//
// After running it, open the agent in the dashboard and click PUBLISH
// (drafts don't go live — see CLAUDE.md gotchas).
//
// Usage:
//   EL_API_KEY=sk_xxx EL_AGENT_ID=agent_xxx node register-elevenlabs-tools.mjs
//
// Or pass them inline:
//   node register-elevenlabs-tools.mjs <api_key> <agent_id>
//
// Defaults: agent ID falls back to the AI Mix Masters production agent
// (agent_2601kqm4g7txfsvv0pkvpe02389p) if EL_AGENT_ID isn't set.
//
// Idempotency: this script CREATES tools every run. If you re-run it, you'll
// get a new set of IDs and the agent will be patched to use those instead —
// the old tool entries are left orphaned in your workspace. Clean those up
// from the dashboard's Tools list if it bothers you.

import { readFileSync, existsSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, resolve } from 'path';

const __dirname = dirname(fileURLToPath(import.meta.url));

const apiKey = process.env.EL_API_KEY || process.argv[2];
const agentId = process.env.EL_AGENT_ID || process.argv[3] || 'agent_2601kqm4g7txfsvv0pkvpe02389p';

if (!apiKey){
  console.error('Missing API key. Set EL_API_KEY=sk_... or pass it as the first arg.');
  console.error('Usage: EL_API_KEY=sk_xxx node register-elevenlabs-tools.mjs');
  process.exit(1);
}

const jsonPath = resolve(__dirname, 'elevenlabs-client-tools.json');
if (!existsSync(jsonPath)){
  console.error('elevenlabs-client-tools.json not found alongside this script.');
  console.error('Path checked:', jsonPath);
  process.exit(1);
}

const tools = JSON.parse(readFileSync(jsonPath, 'utf8'));
console.log(`Loaded ${tools.length} tools from elevenlabs-client-tools.json`);
console.log(`Target agent: ${agentId}`);
console.log('');

const headers = {
  'xi-api-key': apiKey,
  'Content-Type': 'application/json'
};
const base = 'https://api.elevenlabs.io/v1/convai';

// 1. Create each tool and collect the returned ID.
const toolIds = [];
for (let i = 0; i < tools.length; i++){
  const t = tools[i];
  const body = {
    tool_config: {
      type: 'client',
      name: t.name,
      description: t.description,
      // Client-tool variant: parameters is the JSON Schema describing args.
      // expects_response = true so Hope waits for our handler to return and
      // appends the result to the conversation context (per ElevenLabs docs).
      parameters: t.parameters,
      expects_response: true,
      response_timeout_secs: 20
    }
  };
  process.stdout.write(`  [${(i+1).toString().padStart(2)}/${tools.length}] ${t.name.padEnd(28)} → `);
  let res;
  try {
    res = await fetch(`${base}/tools`, {method:'POST', headers, body: JSON.stringify(body)});
  } catch (e){
    console.error(`network error: ${e.message}`);
    process.exit(1);
  }
  if (!res.ok){
    const txt = await res.text().catch(()=>'(no body)');
    console.error(`FAILED ${res.status}`);
    console.error('  body:', txt.slice(0, 500));
    console.error('');
    console.error('If the error mentions an unexpected field, the API schema may have moved.');
    console.error('Last sent payload:', JSON.stringify(body, null, 2).slice(0, 800));
    process.exit(1);
  }
  const j = await res.json();
  if (!j.id){
    console.error(`response missing id:`, JSON.stringify(j).slice(0,300));
    process.exit(1);
  }
  console.log(`id ${j.id}`);
  toolIds.push(j.id);
}

console.log('');
console.log(`Attaching ${toolIds.length} tool IDs to agent ${agentId} …`);

// 2. Patch the agent. Per the API migration docs, we set
// conversation_config.agent.prompt.tool_ids (the legacy `tools` field has
// been removed since July 2025).
const patchBody = {
  conversation_config: {
    agent: {
      prompt: {
        tool_ids: toolIds
      }
    }
  }
};
let pr;
try {
  pr = await fetch(`${base}/agents/${agentId}`, {method:'PATCH', headers, body: JSON.stringify(patchBody)});
} catch (e){
  console.error(`network error patching agent: ${e.message}`);
  process.exit(1);
}
if (!pr.ok){
  const txt = await pr.text().catch(()=>'(no body)');
  console.error(`PATCH agent FAILED ${pr.status}`);
  console.error('  body:', txt.slice(0, 800));
  console.error('  payload:', JSON.stringify(patchBody, null, 2).slice(0, 800));
  process.exit(1);
}

console.log('');
console.log('✓ Done. Open the agent dashboard and click PUBLISH (top-right) to make these tools live.');
console.log(`  https://elevenlabs.io/app/agents/agents/${agentId}`);

File: register_elevenlabs_tools.py
#!/usr/bin/env python3
"""
register_elevenlabs_tools.py

One-shot script to bulk-register every client tool from
elevenlabs-client-tools.json with the ElevenLabs Convai API and attach the
resulting tool IDs to the AI Mix Masters agent's prompt.tool_ids list.

Why: pasting 25 tools one by one in the dashboard form is brutal.

Pure Python, stdlib only — runs on macOS's built-in python3 with no install.

Usage:
    EL_API_KEY=sk_xxx python3 register_elevenlabs_tools.py

Or pass them inline:
    python3 register_elevenlabs_tools.py <api_key> [agent_id]

The agent ID defaults to the AI Mix Masters production agent
(agent_2601kqm4g7txfsvv0pkvpe02389p) if EL_AGENT_ID isn't set.

After running, open the agent dashboard and click PUBLISH (top-right) — drafts
do not go live until published. (CLAUDE.md gotcha — burned an hour on this.)

Idempotency: this script CREATES tools every run. Re-running gets a fresh set
of IDs and the agent is patched to use the new ones; the old tool entries are
left orphaned in your workspace and can be cleaned up from the dashboard's
Tools list.
"""

import os
import sys
import json
import urllib.request
import urllib.error
from pathlib import Path

API_KEY = os.environ.get('EL_API_KEY') or (sys.argv[1] if len(sys.argv) > 1 else None)
AGENT_ID = (
    os.environ.get('EL_AGENT_ID')
    or (sys.argv[2] if len(sys.argv) > 2 else None)
    or 'agent_2601kqm4g7txfsvv0pkvpe02389p'
)

if not API_KEY:
    print('Missing API key. Set EL_API_KEY=sk_... or pass it as the first arg.')
    print('Get one at https://elevenlabs.io/app/settings/api-keys')
    sys.exit(1)

if API_KEY == 'sk_your_key_here':
    print('"sk_your_key_here" is the placeholder text — replace it with your real key.')
    print('Get yours at https://elevenlabs.io/app/settings/api-keys')
    sys.exit(1)

here = Path(__file__).resolve().parent
json_path = here / 'elevenlabs-client-tools.json'
if not json_path.exists():
    print(f'elevenlabs-client-tools.json not found at {json_path}')
    sys.exit(1)

tools = json.loads(json_path.read_text())
print(f'Loaded {len(tools)} tools from elevenlabs-client-tools.json')
print(f'Target agent: {AGENT_ID}')
print()

BASE = 'https://api.elevenlabs.io/v1/convai'
HEADERS = {
    'xi-api-key': API_KEY,
    'Content-Type': 'application/json'
}


def call(method, url, body):
    """POST/PATCH JSON, return (status_code, body_as_dict_or_text)."""
    data = json.dumps(body).encode('utf-8')
    req = urllib.request.Request(url, data=data, headers=HEADERS, method=method)
    try:
        with urllib.request.urlopen(req, timeout=30) as r:
            txt = r.read().decode('utf-8')
            try:
                return r.status, json.loads(txt)
            except json.JSONDecodeError:
                return r.status, txt
    except urllib.error.HTTPError as e:
        txt = e.read().decode('utf-8', errors='replace')
        return e.code, txt
    except urllib.error.URLError as e:
        return 0, f'network error: {e.reason}'


def ensure_param_descriptions(schema, name_hint=None):
    """ElevenLabs requires every leaf parameter (string/number/integer/boolean
    properties AND array items) to set one of: description, dynamic_variable,
    is_system_provided, or constant_value. Our TOOL_DEFS in index.html omit
    description on enum-only fields and array-item types, so we inject a
    sensible fallback before POSTing. Mutates in place."""
    if not isinstance(schema, dict):
        return schema
    typ = schema.get('type')
    # The root parameters object (type=object) is a container — only leaf
    # schemas (string/number/integer/boolean/array) need descriptions.
    if typ and typ != 'object':
        has_description_field = any(
            k in schema for k in ('description', 'dynamic_variable',
                                  'is_system_provided', 'constant_value')
        )
        if not has_description_field:
            schema['description'] = name_hint or typ
    # Recurse into properties.
    props = schema.get('properties')
    if isinstance(props, dict):
        for name, p in props.items():
            ensure_param_descriptions(p, name.replace('_', ' ').strip())
    # Recurse into array items.
    items = schema.get('items')
    if isinstance(items, dict):
        item_hint = (name_hint + ' item') if name_hint else 'item'
        ensure_param_descriptions(items, item_hint)
    return schema


# 1. Create each tool, collect returned IDs.
tool_ids = []
for i, t in enumerate(tools):
    # Deep-copy params before mutating (ensure_param_descriptions adds defaults).
    params = json.loads(json.dumps(t['parameters']))
    ensure_param_descriptions(params)
    body = {
        'tool_config': {
            'type': 'client',
            'name': t['name'],
            'description': t['description'],
            # Client-tool variant: parameters is the JSON Schema describing args.
            # expects_response = True so Hope waits for our handler to return
            # and appends the result to the conversation context.
            'parameters': params,
            'expects_response': True,
            'response_timeout_secs': 20,
        }
    }
    label = f'  [{i+1:2d}/{len(tools)}] {t["name"]:<28} -> '
    print(label, end='', flush=True)
    status, resp = call('POST', f'{BASE}/tools', body)
    if status >= 400 or status == 0:
        print(f'FAILED {status}')
        if isinstance(resp, str):
            print(f'  body: {resp[:600]}')
        else:
            print(f'  body: {json.dumps(resp)[:600]}')
        print()
        print('If the error mentions an unexpected field, the API schema may have shifted.')
        print('Last sent payload:')
        print(json.dumps(body, indent=2)[:1200])
        sys.exit(1)
    if not isinstance(resp, dict) or 'id' not in resp:
        print(f'response missing id: {str(resp)[:300]}')
        sys.exit(1)
    print(f'id {resp["id"]}')
    tool_ids.append(resp['id'])

print()
print(f'Attaching {len(tool_ids)} tool IDs to agent {AGENT_ID} ...')

# 2. Patch the agent. Per the API migration docs, set
# conversation_config.agent.prompt.tool_ids — the legacy `tools` field has
# been removed since July 2025.
patch_body = {
    'conversation_config': {
        'agent': {
            'prompt': {
                'tool_ids': tool_ids
            }
        }
    }
}
status, resp = call('PATCH', f'{BASE}/agents/{AGENT_ID}', patch_body)
if status >= 400 or status == 0:
    print(f'PATCH agent FAILED {status}')
    if isinstance(resp, str):
        print(f'  body: {resp[:800]}')
    else:
        print(f'  body: {json.dumps(resp)[:800]}')
    sys.exit(1)

print()
print('Done. Open the agent dashboard and click PUBLISH (top-right) to make the tools live:')
print(f'  https://elevenlabs.io/app/agents/agents/{AGENT_ID}')
